### PR TITLE
Bucky/mempool txsmap

### PR DESCRIPTION
**Pull request by ebuchman**
Addressing #3509

* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by ebuchman** [comment]
Tagged as v0.31.2-rc0. Forgot to bump version in the code though. 

Not sure what's going on with the linter here:

```
--> Running linter
mempool/cache_test.go:22: G404: Use of weak random number generator (math/rand instead of crypto/rand) (gosec)
		rand.Read(txBytes)
Makefile:248: recipe for target 'lint' failed
make: *** [lint] Error 1
Exited with code 2
```

... it is using `crypto/rand` ...

**Comment by Liamsi** [comment]
Regarding the linter, this seems related: https://github.com/securego/gosec/issues/291

File: mempool/mempool.go [hunk 1]
```diff
@@ -149,6 +149,11 @@ func TxID(tx []byte) string {
 	return fmt.Sprintf("%X", types.Tx(tx).Hash())
 }
 
+// txKey is the fixed length array sha256 hash used as the key in maps.
+func txKey(tx types.Tx) [sha256.Size]byte {
```
Comment (Liamsi) [inline mempool/mempool.go]: This little helper makes this much easier to read 👍

File: abci/client/socket_client.go [hunk 2]
```diff
@@ -86,6 +87,7 @@ func (cli *socketClient) OnStop() {
 	cli.mtx.Lock()
 	defer cli.mtx.Unlock()
 	if cli.conn != nil {
+		// does this really need a mutex?
```
Comment (melekes) [inline abci/client/socket_client.go]: probably not

File: abci/client/socket_client.go [hunk 3]
```diff
@@ -207,12 +209,15 @@ func (cli *socketClient) didRecvResponse(res *types.Response) error {
 	reqres.Done()            // Release waiters
 	cli.reqSent.Remove(next) // Pop first item from linked list
 
-	// Notify reqRes listener if set
+	// Notify reqRes listener if set (request specific callback).
+	// NOTE: it is possible this callback isn't set on the reqres object
+	// at this point, in which case it will be called after, when it is set.
+	// TODO: should we move this after the resCb call so the order is always consistent?
```
Comment (melekes) [inline abci/client/socket_client.go]: consistent with a localclient, you mean?

File: mempool/mempool.go [hunk 4]
```diff
@@ -446,35 +449,53 @@ func (mem *Mempool) resCb(req *abci.Request, res *abci.Response) {
...
+//
+// External callers of CheckTx, like the RPC, can also pass an externalCb through here that is called
+// when all other response processing is complete.
 //
 // Used in CheckTxWithInfo to record PeerID who sent us the tx.
-func (mem *Mempool) reqResCb(tx []byte, peerID uint16) func(res *abci.Response) {
+func (mem *Mempool) reqResCb(tx []byte, peerID uint16, externalCb func(*abci.Response)) func(res *abci.Response) {
 	return func(res *abci.Response) {
 		if mem.recheckCursor != nil {
+			// this should not be possible.
```
Comment (melekes) [inline mempool/mempool.go]: should we add a panic? or remove condition?

File: abci/client/socket_client.go (hunk 3 above)
Comment (ebuchman) [reply abci/client/socket_client.go]: Yes exactly. For remote client the reqResCb happens first, for local client it happens second

File: mempool/mempool.go (hunk 4 above)
Comment (ebuchman) [reply mempool/mempool.go]: I think we can remove condition

File: abci/client/socket_client.go (hunk 3 above)
Comment (ebuchman) [reply abci/client/socket_client.go]: follow up in https://github.com/tendermint/tendermint/issues/3513

[the remark this answers is not in the document]
Comment (ebuchman) [reply abci/client/socket_client.go]: folow up in https://github.com/tendermint/tendermint/issues/3513

File: abci/client/socket_client.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebuchman) [reply abci/client/socket_client.go]: follow up in https://github.com/tendermint/tendermint/issues/3513

File: mempool/mempool.go (hunk 4 above)
Comment (ebuchman) [reply mempool/mempool.go]: ok lets panic

File: CHANGELOG_PENDING.md [hunk 5]
```diff
@@ -19,6 +22,8 @@
 
 ### IMPROVEMENTS:
 
-- [CircleCI] \#3497 Move release management to CircleCI
+- [circle] \#3497 Move release management to CircleCI
```
Comment (Liamsi) [inline CHANGELOG_PENDING.md]: This is not clear in the release process documentation. Specifically, how do we name changes in the top-level / root package directory and in other non-golang directories, e.g. `DOCKER`, `.circleci`, etc.
ref: https://github.com/tendermint/tendermint/pull/3503

**Pull request merged**